edk2toollib/uefi/telem/parser_main.py

```diff
@@ -0,0 +1,660 @@
+d           double                  float           8
+s           char[]                  bytes
+p           char[]                  bytes
+P           void *                  integer
+"""
+
+CPER_HEADER_SIZE            = 128 # A CPER header is 128 bytes 
+CPER_SECTION_HEADER_SIZE    = 72  # A CPER section header is 72 bytes
+FriendlyNames               = {}  # dict loaded from friendlynames.csv used to associate guids with friendly strings
+Parsers                     = []  # list of plugin classes which inherit from CPER_SECTION_DATA and are therefore capable of parsing section data
+
+# Using these lambda concat statements to guarantee linear runtime when concatonating strings, and for simplicity
+concat_s = lambda x,y  : (''.join([x,y," "]))  # Concatenate two strings and put a space at the end
+concat_n = lambda x,y  : (''.join([x,y,"\n"])) # Concatenate two strings and put a newline at the end
+concat   = lambda x,y  : (''.join([x,y]))      # Concatenate two strings
+
+# Useful lambdas
+empty    = lambda x     : x == "" or x == None or x == []   # True if x is an empty string, list, etc.
+neq      = lambda x,y   : x != y                            # True if x and y are not equal
+eq       = lambda x,y   : x == y                            # True if x and y are equal
+
+class CPER(object):
+
+    def __init__(self, input):
+        self.RawData = bytearray.fromhex(input)
+        self.Header = None
+        self.SetCPERHeader()
+        self.SectionHeaders = []
+        self.SetSectionHeaders()
+
+    ##
+    # Turn the portion of the raw input associated with the CPER head into a CPER_HEAD object
+    ##
+    def SetCPERHeader(self):
+        temp = self.RawData[:CPER_HEADER_SIZE]
+        self.Header = CPER_HEADER(temp)
+
+    ##
+    # Set each of the section headers to CPER_SECTION_HEADER objects
+    ##
+    def SetSectionHeaders(self):
+        # Section of RawData containing the section headers
+        temp = self.RawData[CPER_HEADER_SIZE:CPER_HEADER_SIZE + (CPER_SECTION_HEADER_SIZE * self.Header.SectionCount)]
+
+        # Go through each section header and attempt to create a CPER_SECTION_HEADER object. Store each header in SectionHeaders list
+        for x in range(self.Header.SectionCount):
+            try: # Don't want to stop runtime if parsing fails
+                self.SectionHeaders.append(CPER_SECTION_HEADER(temp[x * CPER_SECTION_HEADER_SIZE: (x + 1) * CPER_SECTION_HEADER_SIZE]))
+            except:
+                print("Error parsing section header %d" % x)
+
+    ##
+    # Get each of the actual section data pieces and either pass it to something which can parse it, or dump the hex
+    ##
+    def ParseSectionData(self, s):
+        p = CheckPluginsForGuid(s.SectionType) # Runs CanParse() on each plugin in .\plugins\ folder to see if it can parse the section type 
+        
+        # if we've found a plugin which can parse this section
+        if p != None: 
```

Comment:
should we try multiple plugins if we find multiple plugins that support this guid?

Reply:
My idea was that the first one found wins. There's no harm in doing it another way, though.